File: Data Structures & Algorithms/course-schedule-ii/submission-5.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:

        preMap = [[] for i in range(numCourses)]

        for course, pre in prerequisites:
            preMap[course].append(pre)

        path  = []
        visiting = set()

        def dfs(n):
            if n in visiting:
                return False
            if preMap[n] == []:
                if n not in path:
                    path.append(n)
                return True
            
            visiting.add(n)

            for pre in preMap[n]:
                if not dfs(pre):
                    return False
            
            visiting.remove(n)
            preMap[n] = []
            path.append(n)
            return True
        for i in range(numCourses):
            if not dfs(i):
                return []

        return path

File: Data Structures & Algorithms/course-schedule/submission-5.java
class Solution {
        HashMap<Integer, ArrayList<Integer>> preMap;
        HashSet<Integer> visited;
        HashSet<Integer> visiting;
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        preMap = new HashMap<>();
        visited = new HashSet<>();
        visiting = new HashSet<>();
        for (int[] prereq : prerequisites) {
            int pre = prereq[1];
            int course = prereq[0];
            if (!preMap.containsKey(course)) {
                preMap.put(course, new ArrayList<>());
            }

            preMap.get(course).add(pre);
        }

        for (int n = 0; n < numCourses; n++) {
            if (!dfs(n)) return false;
        }

        return true;
        
    }

    private boolean dfs(int n) {
        if (visited.contains(n)) return true;
        if (visiting.contains(n)) return false;

        visiting.add(n);

        for (int pre : preMap.getOrDefault(n, new ArrayList<>())) {
            if (!dfs(pre)) return false;
        }
        visiting.remove(n);
        visited.add(n);
        return true;
    }
}
